Async RPC server support

zhengrenjie opened the pull request:
An auto-detected Async-RPC. 
If a business method's return-type is CompletableFuture, framework will treat it as a async-method, both client and server. 
In other cases, the framework treat it as a sync-method. 
It is only different from the business's point of view, when inside the framework is always async no matter what the business-method's mode(sync or async) is.

Commit:
init for async server support

Commit:
delete files

jovany-wang commented on common/src/main/java/org/dst/drpc/codec/Codec.java:
```diff
@@ -0,0 +1,14 @@
+import org.dst.drpc.exception.CodecException;
+
+/**
+ * 编解码器
```
Comment In English plz.

jovany-wang commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+  /**
+   * 协议
+   */
+  private String protocol;
```
This should be `enum`.

jovany-wang commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+  /**
+   * 服务端口
+   */
+  private int port;
```
No need use the 2 fields `host` and `port`, just use `address` instead of them.

jovany-wang commented on common/src/main/java/org/dst/drpc/common/Void.java:
```diff
@@ -0,0 +1,17 @@
+package org.dst.drpc.common;
+
+
+public class Void {
```
any usage?

senyer replied:
The same question

jovany-wang commented on common/src/main/java/org/dst/drpc/constants/CodecConstants.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  enum DataType {
+    VOID((byte) 0x00),
```
```suggestion
    NONE((byte) 0x00),
```

jovany-wang commented on common/src/main/java/org/dst/drpc/constants/ParamConstants.java:
```diff
@@ -0,0 +1,12 @@
+
+  String serialization = "Serialization";
+
+  String isAsync = "IsAsync";
```
boolean

jovany-wang commented on common/src/main/java/org/dst/drpc/exception/DstException.java:
```diff
@@ -0,0 +1,25 @@
+package org.dst.drpc.exception;
+
+
+public class DstException extends RuntimeException {
```
```suggestion
public class DrpcException extends RuntimeException {
```

jovany-wang commented on common/src/main/java/org/dst/drpc/model/IpPortPair.java:
```diff
@@ -0,0 +1,46 @@
+import java.util.Objects;
+
+
+public class IpPortPair {
```
```suggestion
public class DrpcAddress {
```

jovany-wang commented on common/src/main/java/org/dst/drpc/utils/ReflectUtils.java:
```diff
@@ -0,0 +1,109 @@
+    List<Method> ret = new ArrayList<>();
+    for (Method method : methods) {
+      boolean isPublic = Modifier.isPublic(method.getModifiers());
+      boolean isNotObjectClass = method.getDeclaringClass() != Object.class;
```
```suggestion
      final boolean isNotObjectClass = method.getDeclaringClass() != Object.class;
```

jovany-wang commented on common/src/main/java/org/dst/drpc/utils/RequestIdGenerator.java:
```diff
@@ -0,0 +1,33 @@
+public class RequestIdGenerator {
+
+  private static LongAdder curId = new LongAdder();
+  private static long MAX_PER_ROUND = 1 << 24;
```
final

jovany-wang commented on common/src/main/java/org/dst/drpc/utils/RequestIdGenerator.java:
```diff
@@ -0,0 +1,33 @@
+    return curId.longValue();
+  }
+
+  public static void main(String[] args) {
```
What's the purpose of this `main` method?

senyer commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+      throw new IllegalArgumentException("URL: address error: " + address, e);
+    }
+  }
+  /* ------ */
```
remove this？

senyer commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+  }
+  /* ------ */
+
+  /* static method */
```
remove this？

senyer commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+  }
+
+  /**
+   * copy from motan, xixi
```
xixi is a name?
maybe you can put a link?

jovany-wang replied:
I think this comment should be removed for such project.

senyer commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+   */
+  public static URL parse(String url) {
+    if (url == null || url.length() == 0) {
+      throw new DstException("URL: url is null");
```
DrpcException

senyer replied:
How about URLException("null") or null Exception?

anyway , i think we should define  exception type in string

senyer commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,364 @@
+
+
+
+  /* -- 获取配置方法 -- */
```
in English

senyer commented on common/src/main/java/org/dst/drpc/constants/CodecConstants.java:
```diff
@@ -0,0 +1,67 @@
+
+  short MAGIC_HEAD = 9527;
+
+  int HEADER_LENGTH = 16/* bytes, not bits */;
```
replace this commont?
How about directly describe the meaning of this attribute

senyer commented on common/src/main/java/org/dst/drpc/constants/CodecConstants.java:
```diff
@@ -0,0 +1,67 @@
+
+  enum Version {
+    VERSION_1((byte) 0x01),
+
```
Why blank line?

senyer commented on common/src/main/java/org/dst/drpc/constants/ParamConstants.java:
```diff
@@ -0,0 +1,12 @@
+
+  String transportType = "TransportType";
+
+  String serialization = "Serialization";
```
any usage?

senyer commented on common/src/main/java/org/dst/drpc/utils/NetUtils.java:
```diff
@@ -0,0 +1,177 @@
+
+  /**
+   * <pre>
+   * 查找策略：首先看是否已经查到ip --> hostname对应的ip --> 根据连接目标端口得到的本地ip --> 轮询网卡
```
In English

senyer commented on common/src/main/java/org/dst/drpc/utils/NetUtils.java:
```diff
@@ -0,0 +1,177 @@
+          try {
+            socket.close();
+          } catch (Throwable e) {
+            // ignore
```
why ignore?

senyer commented on common/src/main/java/org/dst/drpc/utils/RequestIdGenerator.java:
```diff
@@ -0,0 +1,33 @@
+  private static long MAX_PER_ROUND = 1 << 24;
+
+  static {
+    curId.increment(); // 从1开始
```
in english

senyer commented on example/src/main/java/org/dst/drpc/async/Client.java:
```diff
@@ -0,0 +1,40 @@
+    reference.setInterfaceClass(IServer.class);
+
+    IServer server = reference.getReference();
+//    System.out.println(server.say());
```
remove this

senyer commented on example/src/main/java/org/dst/drpc/async/IServer.java:
```diff
@@ -0,0 +1,14 @@
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remove this commot

senyer commented on example/src/main/java/org/dst/drpc/async/IServerImpl.java:
```diff
@@ -0,0 +1,36 @@
+
+  @Override
+  public String say() {
+    return "dst";
```
As a standalone rpc module, do we need to introduce the semantics of `dst`?

senyer commented on example/pom.xml:
```diff
@@ -0,0 +1,22 @@
+    </dependency>
+  </dependencies>
+
+</project>
```
blank line

senyer commented on example/src/main/java/org/dst/drpc/async/IServerImpl.java:
```diff
@@ -0,0 +1,36 @@
+    try {
+      Thread.sleep(t);
+    } catch (Exception e) {
+      // ignore
```
why ignore

senyer commented on example/src/main/java/org/dst/drpc/async/Server.java:
```diff
@@ -0,0 +1,21 @@
+import org.dst.drpc.Exporter;
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remote this commont

senyer commented on protocol/src/main/java/org/dst/drpc/DefaultInvoker.java:
```diff
@@ -0,0 +1,35 @@
+import org.dst.drpc.common.URL;
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remove this

senyer commented on protocol/src/main/java/org/dst/drpc/DefaultInvoker.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Override
+  public URL getURL() {
+    return null;
```
why return null?

senyer commented on protocol/src/main/java/org/dst/drpc/DefaultInvoker.java:
```diff
@@ -0,0 +1,35 @@
+
+  @Override
+  public Class<T> getInterface() {
+    return null;
```
why null?

senyer commented on protocol/src/main/java/org/dst/drpc/Exporter.java:
```diff
@@ -0,0 +1,59 @@
+
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remove this

senyer commented on protocol/src/main/java/org/dst/drpc/Invoker.java:
```diff
@@ -0,0 +1,19 @@
+
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remove

senyer commented on protocol/src/main/java/org/dst/drpc/Reference.java:
```diff
@@ -0,0 +1,55 @@
+import org.dst.drpc.proxy.ProxyFactory;
+
+/**
+ * @author zrj CreateDate: 2019/10/28
```
remove

senyer commented on protocol/src/main/java/org/dst/drpc/Reference.java:
```diff
@@ -0,0 +1,55 @@
+    serverUrl.setAddress(address.substring(address.indexOf("://") + "://".length()));
+  }
+
+  // fixme : 不要设置，自动根据返回类型自行判断
```
english

senyer commented on protocol/src/main/java/org/dst/drpc/ServerImpl.java:
```diff
@@ -0,0 +1,82 @@
+  private URL serverUrl;
+
+  /**
+   * 找到所有interfaceClazz可以调用的方法，并缓存下来，缓存名字要保留参数类型的完整名称，防止函数重载
```
english

senyer commented on protocol/src/main/java/org/dst/drpc/ServerImpl.java:
```diff
@@ -0,0 +1,82 @@
+   */
+  public ServerImpl(URL serverUrl, T ref, Class<T> interfaceClazz) {
+    if (!interfaceClazz.isInterface()) {
+      throw new DstException("ServerImpl: interfaceClazz is not a interface!");
```
i think we shouldn't define Exception in a string description.
maybe coustom exception is a good chioce

senyer commented on protocol/src/main/java/org/dst/drpc/ServerImpl.java:
```diff
@@ -0,0 +1,82 @@
+/**
+ * @author zrj CreateDate: 2019/10/28
+ */
+public class ServerImpl<T> implements Invoker<T> {
```
why called ServerImpl ,but not implements a Server interface?

senyer commented on protocol/src/main/java/org/dst/drpc/ServerImpl.java:
```diff
@@ -0,0 +1,82 @@
+    String methodName = ReflectUtils.getMethodDesc(request.getMethodName(), request.getArgsType());
+    Method method = methodMap.get(methodName);
+    if (method == null) {
+      response.setThrowable(new DstException("ServerImpl: can't find method: " + methodName));
```
the same

senyer commented on protocol/src/main/java/org/dst/drpc/proxy/ProxyFactory.java:
```diff
@@ -0,0 +1,16 @@
+
+public class ProxyFactory<T> {
+
+  @SuppressWarnings("unchecked")
```
Is that suitable use @SuppressWarnings?

@jovany-wang

senyer commented on protocol/src/main/java/org/dst/drpc/proxy/ProxyHandler.java:
```diff
@@ -0,0 +1,115 @@
+  @Override
+  public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+    if (isLocalMethod(method)) {
```
Maybe we can encapsulate this part of the code into a method

senyer replied:
handleLocalMethod()

senyer commented on protocol/src/main/java/org/dst/drpc/proxy/ProxyHandler.java:
```diff
@@ -0,0 +1,115 @@
+      return future;
+    }
+
+    // 如果不是异步的，直接同步返回结果。
```
english

senyer commented on transport/pom.xml:
```diff
@@ -0,0 +1,38 @@
+    <dependency>
+      <groupId>io.netty</groupId>
+      <artifactId>netty-all</artifactId>
+      <version>4.1.16.Final</version>
```
maybe this version is a little outdate?

senyer commented on transport/pom.xml:
```diff
@@ -0,0 +1,38 @@
+  </dependencies>
+
+
+</project>
```
blank line

senyer commented on transport/src/main/java/org/dst/drpc/api/AbstractChannel.java:
```diff
@@ -0,0 +1,40 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ *
```
no comment

senyer commented on transport/src/main/java/org/dst/drpc/api/AbstractClient.java:
```diff
@@ -0,0 +1,69 @@
+
+  private static final int NEW = 0;
+  private static final int CONNECTED = 1;
+  private static final int DISCONNECTED = 2;
```
Can you introduce the design of this code？

senyer commented on transport/src/main/java/org/dst/drpc/api/AbstractServer.java:
```diff
@@ -0,0 +1,80 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * 服务端
```
introduce detail in english or remove this？

senyer commented on transport/src/main/java/org/dst/drpc/api/AbstractServer.java:
```diff
@@ -0,0 +1,80 @@
+  private Codec codec;
+
+  /**
+   * Server只要存在就不为null
```
english

senyer commented on transport/src/main/java/org/dst/drpc/api/AbstractServer.java:
```diff
@@ -0,0 +1,80 @@
+    /*
+      fixme 这是一个简单的线程池实现，不阻塞网络io线程，这里就是不阻塞netty的io线程，线程数应该要大一点，dubbo默认是200，这个参数需要从url读取
+     */
+    executor = Executors.newFixedThreadPool(GlobalConstants.threadNumber * 2);
```
I think we need to study the number of thread pools more deeply to get a better choice

zhengrenjie replied:
yep, this parameters will be configurable a few version later.

senyer commented on transport/src/main/java/org/dst/drpc/api/Channel.java:
```diff
@@ -0,0 +1,54 @@
+  /**
+   * 检查一个信道是否打开。
+   */
+//  boolean isOpen();
```
remove this

senyer commented on transport/src/main/java/org/dst/drpc/api/Channel.java:
```diff
@@ -0,0 +1,54 @@
+  /**
+   * has attribute.
+   *
+   * @param key key.
```
I don't think it's necessary to add these comments. The semantics of method names are very clear

senyer commented on transport/src/main/java/org/dst/drpc/api/DefaultRoutableHandler.java:
```diff
@@ -0,0 +1,71 @@
+
+
+/**
+ *
```
remove

senyer commented on transport/src/main/java/org/dst/drpc/api/DefaultRoutableHandler.java:
```diff
@@ -0,0 +1,71 @@
+
+  @Override
+  public Handler getHandlerByServerName(URL url) {
+
```
Why null?

senyer commented on transport/src/main/java/org/dst/drpc/api/DefaultRoutableHandler.java:
```diff
@@ -0,0 +1,71 @@
+  }
+
+  @Override
+  public void merge(RoutableHandler handler) {
```
How about add some commont to introduce your design?

senyer replied:
I found it in interface

senyer commented on transport/src/main/java/org/dst/drpc/api/Handler.java:
```diff
@@ -0,0 +1,19 @@
+public interface Handler {
+
+  /**
+   * 标志这个Handler服务的服务
```
english

senyer commented on transport/src/main/java/org/dst/drpc/api/Server.java:
```diff
@@ -0,0 +1,26 @@
+import org.dst.drpc.common.URL;
+
+/**
+ *
```
empty

senyer commented on transport/src/main/java/org/dst/drpc/api/Server.java:
```diff
@@ -0,0 +1,26 @@
+
+  RoutableHandler getRoutableHandler();
+
+  Executor getExecutor();
```
why getExecutor（）

senyer commented on transport/src/main/java/org/dst/drpc/api/ServerFactory.java:
```diff
@@ -0,0 +1,50 @@
+ * 必须是单例的
+ */
+public abstract class ServerFactory {
+
```
english

senyer commented on transport/src/main/java/org/dst/drpc/api/async/AsyncResponse.java:
```diff
@@ -0,0 +1,19 @@
+/**
+ * @author zrj CreateDate: 2019/10/29
+ */
+public interface AsyncResponse extends Response, CompletionStage<Response>, Future<Response> {
```
Coloud you introduce why excends CompletionStage ?

senyer commented on transport/src/main/java/org/dst/drpc/api/async/DefaultAsyncResponse.java:
```diff
@@ -0,0 +1,163 @@
+
+  @Override
+  public boolean complete(Response value) {
+    // 封印这个方法
```
????
could you introduce the propose?

senyer commented on transport/src/main/java/org/dst/drpc/api/async/DefaultAsyncResponse.java:
```diff
@@ -0,0 +1,163 @@
+//    checkDone();
+    try {
+      return get();
+    } catch (Exception e) {
```
Is it necessary to continue encapsulating and catching exceptions

senyer commented on transport/src/main/java/org/dst/drpc/api/async/DefaultResponse.java:
```diff
@@ -0,0 +1,83 @@
+import java.util.Map;
+
+/**
+ * @author zrj CreateDate: 2019/10/29
```
empty

senyer commented on transport/src/main/java/org/dst/drpc/api/async/DefaultResponse.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public class DefaultResponse implements Response {
+
+  public static final String IS_ASYNC = "IS_ASYNC";
```
boolean?

senyer commented on transport/src/main/java/org/dst/drpc/api/async/facade/ResponseCompletableFutureFacade.java:
```diff
@@ -0,0 +1,79 @@
+  public Object getValue() {
+    try {
+      return get();
+    } catch (Exception e) {
```
Is it necessary to continue encapsulating and catching exceptions

senyer commented on transport/src/main/java/org/dst/drpc/api/async/facade/ResponseFutureFacade.java:
```diff
@@ -0,0 +1,45 @@
+/**
+ * @author zrj CreateDate: 2019/11/1
+ */
+public class ResponseFutureFacade implements Future {
```
Why the facade design not used anywhere?

zhengrenjie replied:
It was designed intended to return back to the invoker to hid some details which in the framework inside. But it not used and will be remove in next commit.

senyer commented on transport/src/main/java/org/dst/drpc/codec/DstCodec.java:
```diff
@@ -0,0 +1,279 @@
+/**
+ * 编解码核心类：
+ *
+ * 协议头：
```
english

Commit:
delete some none-use classes, fix some problems

Commit:
delete 'serverUrl' from ServerImpl

Commit:
delete pom.xml blank line and add some annotation

Commit:
add a log

Commit:
change void to none

Commit:
delete none-use feild

zhengrenjie commented:
I have committed some changes according to your comments.
1. deleted some none-use annotations left when developing.
2. deleted some field that not make sense and never be used.
3. deleted all none-use class files.
4. deleted the blank line in pom.xml.
5. added some logging and annotation.

There is remaining a few problems.
1. As I have deleted no useful annotation, the new valuable annotation has not be presented yet.
2. Some problems such as using Custom-Exception instead of String-Exception is not fixed yet, because the design of Exception is not an easy work which may spend a lot of time. Furthermore, as the development goes by, the design of Exception may be changed serval times. So I prefer to use a String-Exception in the early time and improve the Exception-System in later developments. 
Other issues I also prefer to improve them in later developments as they are work well in the examples. If there is something must to be changed or fixed, just tell me.

@jovany-wang @senyer

Commit:
fix

senyer commented:
Great job!！

Commit:
fix

Commit:
invoke 'pos += 2' later

Commit:
fix

Commit:
DstCodec not use static method to transform data

Commit:
There is no need to cast Response to AsyncResponse

MrKingHH commented on protocol/pom.xml:
```diff
@@ -0,0 +1,21 @@
+      <version>1.0-SNAPSHOT</version>
+    </dependency>
+  </dependencies>
+</project>
```
One more line plz.

MrKingHH commented on common/src/main/java/org/dst/drpc/common/URL.java:
```diff
@@ -0,0 +1,337 @@
+    return def;
+  }
+
+  public int getInt(String key, int def) {
+    if (parameters.containsKey(key)) {
+      return Short.valueOf(parameters.get(key));
+    }
+    return def;
+  }
```
The method name `getInt(String key, int def)` is hard to understand here. Can we change another name?

MrKingHH commented on transport/src/main/java/org/dst/drpc/netty/NettyClient.java:
```diff
@@ -0,0 +1,134 @@
+                processResponse(ctx, (Response) object);
+              }
+
+              private void processResponse(ChannelHandlerContext ctx, Response response) {
```
The parameter `ctx` is not used. Can u remove it?

MrKingHH reviewed (changes requested):
Thanks! I have left some comments to be changed.

Commit:
delete methed arg

Commit:
add pb example

Commit:
delete target files

Commit:
pb

Commit:
new blank line

Commit:
new blank line

jovany-wang reviewed (approved):
Let's merge this as a prototype.
